numpy/lib/npyio.py

```diff
@@ -599,13 +611,41 @@ def savez_compressed(file, *args, **kwds):
     """
     _savez(file, args, kwds, True)
 
+@contextlib.contextmanager
+def DeleteOnContextExitNamedTemporaryFile(*args, **kwds):
+    """ Factory for a NamedTemporaryFile that is
+        deleted on context exit but not on file close.
+        
+        Usefull replacement for NamedTemporaryFile,
+        as WinNT prevents a file to be opened twice.
+        
+        Typical case:
+        with DeleteOnExitNamedTemporaryFile() as fid:
+            foo(fid)
+            fid.close()
+            # File not deleted here
+            bar(fid.name)
+        # File deleted here
+    """
+    # Import deferred for startup time improvement
+    import tempfile    
+    local_kwds = kwds.copy()
+    local_kwds['delete'] = False
+    filename = None
+    with tempfile.NamedTemporaryFile(*args, **local_kwds) as fid:
+        filename = fid.name
+        yield fid
+    try:
```

Comment:
The os.close is performed by the context manager from NamedTemporaryFile at the exit of the with statement.

The whole pain came from the fact that NamedTemporaryFile is a factory function not a class, so you can not just subclass it.